File: simter-file-rest-webflux/src/test/java/tech/simter/file/rest/webflux/handler/DownloadFileHandlerTest.java
package tech.simter.file.rest.webflux.handler;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.test.mock.mockito.SpyBean;
import org.springframework.core.io.FileSystemResource;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
import org.springframework.test.web.reactive.server.EntityExchangeResult;
import org.springframework.test.web.reactive.server.WebTestClient;
import reactor.core.publisher.Mono;
import tech.simter.file.po.Attachment;
import tech.simter.file.rest.webflux.WebFluxConfiguration;
import tech.simter.file.service.AttachmentService;

import java.io.IOException;
import java.time.OffsetDateTime;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.springframework.http.MediaType.APPLICATION_OCTET_STREAM_VALUE;

/**
 * Test DownloadFileHandler.
 *
 * @author JF
 */
@SpringJUnitConfig(classes = {
  WebFluxConfiguration.class, DownloadFileHandler.class
})
@MockBean(AttachmentService.class)
@TestPropertySource(properties = "app.file.root=src/test")
class DownloadFileHandlerTest {
  private WebTestClient client;
  @SpyBean
  private DownloadFileHandler handler;
  @Autowired
  private AttachmentService service;
  @Value("${app.file.root}")
  private String fileRootDir;

  @BeforeEach
  void setUp() {
    client = WebTestClient.bindToRouterFunction(handler.router()).build();
  }

  @Test
  void download() throws IOException {
    // mock service return value
    String name = "logback-test";
    String ext = "xml";
    String fileName = name + "." + ext;
    String id = UUID.randomUUID().toString();
    long fileSize = new FileSystemResource(fileRootDir + "/resources/" + fileName).contentLength();
    Attachment attachment = new Attachment(id, "resources/" + name + "." + ext, name, ext, fileSize,
      OffsetDateTime.now(), "Simter");
    Mono<Attachment> expected = Mono.just(attachment);
    when(service.get(id)).thenReturn(expected);

    // invoke request
    EntityExchangeResult<byte[]> result = client.get().uri("/" + id)
      .exchange()
      .expectStatus().isOk()
      .expectHeader().valueEquals("Content-Type", APPLICATION_OCTET_STREAM_VALUE)
      .expectHeader().valueEquals("Content-Length", String.valueOf(fileSize))
      .expectHeader().valueEquals("Content-Disposition", "attachment; filename=\"" + fileName + "\"")
      .expectBody().returnResult();

    // verify response body
    assertNotNull(result.getResponseBody());
    assertEquals(result.getResponseBody().length, fileSize);

    // verify method service.get invoked
    verify(service).get(id);
  }
}

File: simter-file-rest-webflux/src/test/kotlin/tech/simter/file/rest/webflux/handler/DownloadFileHandlerTest.kt
package tech.simter.file.rest.webflux.handler

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import org.mockito.Mockito.`when`
import org.mockito.Mockito.verify
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest
import org.springframework.core.io.FileSystemResource
import org.springframework.http.MediaType.APPLICATION_XML_VALUE
import org.springframework.test.context.TestPropertySource
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig
import org.springframework.test.web.reactive.server.WebTestClient
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono
import tech.simter.exception.PermissionDeniedException
import tech.simter.file.core.AttachmentService
import tech.simter.file.core.domain.Attachment
import tech.simter.file.impl.domain.AttachmentImpl
import tech.simter.file.rest.webflux.UnitTestConfiguration
import java.time.OffsetDateTime
import java.util.*

/**
 * Test [DownloadFileHandler].
 *
 * @author JF
 * @author RJ
 * @author zh
 */
@SpringJUnitConfig(UnitTestConfiguration::class)
@WebFluxTest
@TestPropertySource(properties = ["simter.file.root=src/test"])
class DownloadFileHandlerTest @Autowired constructor(
  private val client: WebTestClient,
  private val service: AttachmentService,
  @Value("\${simter.file.root}") private val fileRootDir: String
) {
  @Test
  fun found() {
    // mock service return value
    val name = "logback-test"
    val ext = "xml"
    val fileName = "$name.$ext"
    val id = UUID.randomUUID().toString()
    val now = OffsetDateTime.now()
    val fileSize = FileSystemResource("$fileRootDir/resources/$fileName").contentLength()
    val attachment: Attachment = AttachmentImpl(id, "$name.$ext", name, ext, fileSize,
      now, "Simter", now, "Simter", "0")
    val expected = Mono.just(attachment)
    `when`(service.get(id)).thenReturn(expected)
    `when`(service.getFullPath(id)).thenReturn("resources/$name.$ext".toMono())

    // invoke request
    val result = client.get().uri("/$id")
      .exchange()
      .expectStatus().isOk
      .expectHeader().valueEquals("Content-Type", APPLICATION_XML_VALUE)
      .expectHeader().valueEquals("Content-Length", fileSize.toString())
      .expectHeader().valueEquals("Content-Disposition", "attachment; filename=\"$fileName\"")
      .expectBody().returnResult()

    // verify response body
    assertNotNull(result.responseBody)
    assertEquals(result.responseBody!!.size.toLong(), fileSize)

    // verify method service.get invoked
    verify(service).get(id)
    verify(service).getFullPath(id)
  }

  @Test
  fun notFound() {
    // mock
    val id = UUID.randomUUID().toString()
    `when`(service.get(id)).thenReturn(Mono.empty())
    `when`(service.getFullPath(id)).thenReturn(Mono.empty())

    // invoke
    client.get().uri("/$id").exchange().expectStatus().isNotFound

    // verify
    verify(service).get(id)
    verify(service).getFullPath(id)
  }

  @Test
  fun failedByPermissionDenied() {
    // mock
    val id = UUID.randomUUID().toString()
    `when`(service.get(id)).thenReturn(Mono.error(PermissionDeniedException()))

    // invoke
    client.get().uri("/$id").exchange().expectStatus().isForbidden

    // verify
    verify(service).get(id)
  }
}